libshaderc_spvc/src/spvc.cc

```diff
@@ -164,74 +174,370 @@ void shaderc_spvc_compiler_release(shaderc_spvc_compiler_t compiler) {
+    return ((Compiler*)user_data)->parse_instruction(inst);
+  }
+
+  // Duplicate behavior of SPIRV-Cross.
+  spv_result_t parse_instruction(const spv_parsed_instruction_t* inst) {
+    switch (inst->opcode) {
+      case spv::OpExtInstImport: {
+        uint32_t id;
+        std::string ext;
+        GetOps(inst, id, ext);
+        if (ext == "GLSL.std.450")
+          set<spirv_cross::SPIRExtension>(id, spirv_cross::SPIRExtension::GLSL);
+        else if (ext == "SPV_AMD_shader_ballot")
+          set<spirv_cross::SPIRExtension>(
+              id, spirv_cross::SPIRExtension::SPV_AMD_shader_ballot);
+        else if (ext == "SPV_AMD_shader_explicit_vertex_parameter")
+          set<spirv_cross::SPIRExtension>(
+              id, spirv_cross::SPIRExtension::
+                      SPV_AMD_shader_explicit_vertex_parameter);
+        else if (ext == "SPV_AMD_shader_trinary_minmax")
+          set<spirv_cross::SPIRExtension>(
+              id, spirv_cross::SPIRExtension::SPV_AMD_shader_trinary_minmax);
+        else if (ext == "SPV_AMD_gcn_shader")
+          set<spirv_cross::SPIRExtension>(
+              id, spirv_cross::SPIRExtension::SPV_AMD_gcn_shader);
+        else
+          set<spirv_cross::SPIRExtension>(
+              id, spirv_cross::SPIRExtension::Unsupported);
+        // Other SPIR-V extensions which have ExtInstrs are currently not
+        // supported.
+      } break;
+
+      case spv::OpEntryPoint: {
+        spv::ExecutionModel model;
+        uint32_t id;
+        std::string name;
+        std::vector<uint32_t> interface_variables;
+        GetOps(inst, model, id, name, interface_variables);
+
+        auto itr = ir.entry_points.insert(
+            std::make_pair(id, spirv_cross::SPIREntryPoint(id, model, name)));
+        auto& e = itr.first->second;
+        std::copy(interface_variables.begin(), interface_variables.end(),
+                  back_inserter(e.interface_variables));
+        ir.set_name(id, e.name);
+        if (!ir.default_entry_point) ir.default_entry_point = id;
+      } break;
+
+      case spv::OpExecutionMode: {
+        uint32_t entry_point;
+        spv::ExecutionMode mode;
+        GetOps(inst, entry_point, mode);
+
+        auto& execution = ir.entry_points[entry_point];
+        execution.flags.set(mode);
+
+        switch (mode) {
+          case spv::ExecutionModeInvocations:
+            GetOpsRecur(inst, 2, execution.invocations);
+            break;
+          case spv::ExecutionModeLocalSize:
+            GetOpsRecur(inst, 2, execution.workgroup_size.x,
+                        execution.workgroup_size.y, execution.workgroup_size.z);
+            break;
+          case spv::ExecutionModeOutputVertices:
+            GetOpsRecur(inst, 2, execution.output_vertices);
+            break;
+          default:
+            break;
+        }
+      } break;
+
+      case spv::OpCapability: {
+        spv::Capability cap;
+        GetOps(inst, cap);
```

Comment:
I very much do not like this non-const ref argument pattern, and the effects of the recursive template thing.

Also the recursive template GetOps construction won't deal well with variably-determined arguments, e.g. for image instructions (see the use of the bitmask word that controls image operands). There are other cases like that too, e.g. loop control, memory access.

Please take a look at SPIRV-Tools source/opt/ir_loader.h and ir_loader.cpp
That is a much more straightforward conversion from parsed instructions (from spvBinarParse) to an internal representation.